### ISSUE TITLE
JetBrains keymap incomplete

### ISSUE THREAD

**Issue by KaKi87** [zed-industries/zed#14639]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

As a regular JetBrains IDE user, I noticed that the JetBrains keymap is incomplete, for example :

- `Ctrl`-`W` & `Ctrl`-`Shift`-`W` don't [extend/shrink selection](https://www.jetbrains.com/help/idea/working-with-source-code.html#editor_code_selection) ;
- `Shift`-`Enter` while searching in file doesn't jump to the previous occurence ;
- `Alt`-`F12` doesn't [toggle terminal](https://www.jetbrains.com/help/idea/terminal-emulator.html) ;
- `Ctrl`-`Shift`-`T` doesn't [open a new terminal tab](https://www.jetbrains.com/help/idea/terminal-emulator.html#new_session) when focused on terminal ;
- `Ctrl`-`Z` directly undoes the last edit instead of first just moving the cursor to the last edited position when not already there.

I've only been trying out Zed for a few hours so there's probably more missing.

### Environment

Zed: v0.143.7 (Zed)
OS: Linux Wayland neon 22.04
Memory: 15.5 GiB
Architecture: x86_64

---

Thanks

PS : Zed is so awesome already, keep it up ! :tada:

**Comment by JosephTLyons** [comment]
I've labeled this as a `good first issue`, but something I will mention is that there is a bit of weirdness with dock toggling. In Zed, we don't have commands to toggle panels (terminal, project, etc), we only have commands to toggle the docks that hold the panels:

- `workspace: toggle left dock`
- `workspace: toggle right dock`
- `workspace: toggle bottom dock`

So we can hook up any key binding to those, but the user may or may not have their terminal in one of those docks. We do have a command to simply focus the terminal, regardless of what dock it is in, and that command will open the dock if its hidden, but it won't toggle it back closed:

- `terminal panel: toggle focus`

So, I'm not entirely sure there's a good way to handle that, unless we just assign the key binding to whatever dock holds the terminal panel by default.

**Comment by KaKi87** [comment]
> we don't have commands to toggle panels (terminal, project, etc)

What's this then ? :thinking: 

![](https://github.com/user-attachments/assets/7c3a6bcb-cdf7-437e-939a-16983015ddb5)

**Comment by notpeter** [comment]
>> we don't have commands to toggle panels (terminal, project, etc)

> What's this then ? 🤔

Our ```ctrl-` ``` behaves differently than VSCode despite the same shortcut by-default.  In VSCode this shortcut will "create terminal, show panel, set focus to terminal" and "hide panel, set focus to editor`.

Basically Joseph is saying that we have shortcuts that toggle the left/right/bottom docks, but that by-default our shortcuts don't toggle the individual panels that may occupy those docks.

**Comment by KaKi87** [comment]
Well, eventually matching the behavior would be much appreciated, but at least changing this shortcut to `Alt`-`F12` as per the JetBrains keymap would still be nice.

Thanks

**Comment by notpeter** [comment]
> Well, eventually matching the behavior would be much appreciated, but at least changing this shortcut now to `Alt`-`F12` as per the JetBrains keymap would still be nice.

Yeah, I've been doing my best to refine the keybindings as best as possible, but there's always room for improvement. Happy to review PRs just tag me, merged one for sublime this morning:

- https://github.com/zed-industries/zed/pull/15089

**Comment by KaKi87** [comment]
> Happy to review PRs just tag me

Okay, where are action strings and key strings documented ?

I found [Key bindings - Zed](https://zed.dev/docs/key-bindings), which lists the default shortcuts, and some examples of action & key strings, but not an exhaustive list :thinking: 

Thanks

**Comment by ken-kuro** [comment]
I've encountered some unexpected keymap behaviors while using Zed. I wasn't able to find comprehensive documentation on Zed's JetBrains base keymap.

Specifically, certain frequently used shortcuts like "Open Terminal" (Alt+F12) and "Format Code" (Ctrl+Alt+L) default to VS Code keybindings, even when the base keymap is set to JetBrains.

I understand this might be a lower priority feature, but any guidance on customizing the keymap or clarifying the base keymap behavior would be greatly appreciated.

**Comment by notpeter** [comment]
@ken-kuro The way things are currently constructed is that Zed has a stacked keymaps:

- default: [default-linux.json](https://github.com/zed-industries/zed/blob/main/assets/keymaps/default-linux.json) which is broadly vscode compatible
- compatibility: [jetbrains.json](https://github.com/zed-industries/zed/blob/main/assets/keymaps/linux/jetbrains.json), [sublime_text.json](https://github.com/zed-industries/zed/blob/main/assets/keymaps/linux/sublime_text.json), etc (optional)
- vim: [vim.json](https://github.com/zed-industries/zed/blob/49eb865e8aaac79f0c33b8b028bf5d47bf733e05/assets/keymaps/vim.json) (optional)
- user: `~/.zed/keymap.json`

Each layer can override the previous, but when using JetBrains we don't specifically unbind all the keys defined in the default (vscode) base keymap.

If you wish to select your compatibility keymap you can spawn "welcome: toggle base keymap selector" from the command palette and select your options.

If there are jetbrains keybinds you use that you find are missing, please comment here. We're always looking to improve and I'm happy to review PRs. Some behaviors (e.g. making cmd-z not undo, but doing a no-op jump to the change first) will require additional code changes, often times as optional parameters for existing actions. For example:
https://github.com/zed-industries/zed/blob/49eb865e8aaac79f0c33b8b028bf5d47bf733e05/assets/keymaps/default-macos.json#L425
While others are just simple additions of new mapping of keys to actions.

**Comment by ken-kuro** [comment]
Regarding the specific example of "toggle terminal," the "Zed way" you described differs significantly from other IDEs like VSCode or JetBrains. In those IDEs, a single keybind typically controls the opening and closing of the terminal panel. However, Zed's approach involves multiple events, such as `workspace::NewTerminal`, `terminal_panel::ToggleFocus` to open, and `pane::CloseActiveItem` in "Terminal" context to close it. This multi-step process is somewhat akin to how Neovim handles its windows (buffers). While I'll propose a solution or submit a pull request when I have one, for now, I think we-users may need to retrain our muscle memory and adapt to Zed's unique approach.

**Comment by KaKi87** [comment]
> Zed's approach involves multiple events, such as `workspace::NewTerminal`, `terminal_panel::ToggleFocus` to open, and `pane::CloseActiveItem` in "Terminal" context to close it [...] we-users may need to retrain our muscle memory and adapt to Zed's unique approach.

Retrain how ? And why ?

**Comment by ken-kuro** [comment]
I’m still getting used to Zed’s philosophy, which seems quite different from other IDEs like JetBrains or VSCode. As someone who primarily uses JetBrains, I don’t have much to add yet, but I agree that the keymap approach takes some adjustment. Personally, I sometimes prefer Vim/Neovim over VSCode or JetBrains for certain tasks or projects—it really depends on what feels right for the situation and gets the job done.

Regarding Zed, if you have any suggestions on how to streamline or improve the current keymap behavior (which I also find challenging), I’d appreciate it if you could share them here. For now, I’m just trying to adapt and get comfortable with it.

**Comment by dam5s** [comment]
In the meantime I have been updating the defaults to match something closer to what I'm used in IntelliJ and with fixes for jumping to search results and such. Not sure if that helps anyone but here is what I have so far -

```  {
    "bindings": {
      "alt-f12": "terminal_panel::ToggleFocus",
      "alt-enter": "editor::OpenExcerpts",
      "cmd-shift-f12": "workspace::CloseAllDocks",
      "cmd-e": "file_finder::Toggle"
    }
  },
  {
    "context": "Editor",
    "bindings": {
      "cmd-=": "editor::UnfoldLines"
    }
  }
]
```

**Comment by chklauser** [comment]
The one thing I miss from the JB family of tools and that I struggle to re-construct in zed is:
 * hitting escape (repeatedly, if necessary) should (eventually) return focus to the file editor

## Why is this is important/useful
 * I don't want to have to remember a dedicated different binding for each kind of dock/panel. ("Oh sorry, your focus is on the agent panel right now. To return from that, it's Ctrl-L, not Ctrl-Shift-E")
 * Sometimes I don't even know which panel has the focus. Having a key sequence that blindly "returns the editor to a known state" is extremely important. (The same way escape in vim always returns you to normal mode)

## Why this is hard to replicate in Zed
  * I can't find an action that just focuses that last file editor pane that had focus
  * You can try to define a custom escape binding for each of the panels to toggle that specific panel. For some reason, I can't get this to work with the AgentPanel, though

**Comment by notpeter** [comment]
> @chklauser: I can't find an action that just focuses that last file editor pane that had focus

This just landed yesterday in Zed Preview: `editor::ToggleFocus`

- https://github.com/zed-industries/zed/pull/34495

@chklauser Can you download [Zed Preview](https://zed.dev/releases/preview) and add this to your Keymap JSON?

```json
  {
    "context": "Dock || Workspace || Terminal || OutlinePanel || ProjectPanel || CollabPanel || (Editor && mode == auto_height)",
    "bindings": { "escape": "editor::ToggleFocus" }
  },
```

**Comment by notpeter** [comment]
I think I've covered everything that was included in this thread in:
- https://github.com/zed-industries/zed/pull/34641

If I missed something or you have additional things you'd like to see added in the JetBrains keymap, please add them here:
- https://github.com/zed-industries/zed/discussions/34643

Thanks all!

**Comment by KaKi87** [comment]
> If I missed something or you have additional things you'd like to see added in the JetBrains keymap, please add them here:
> 
> * [JetBrains Keymap improvements #34643](https://github.com/zed-industries/zed/discussions/34643)

I commented with the items on my initial list that I didn't see in that PR.

**Comment by chklauser** [comment]
> [@chklauser](https://github.com/chklauser) Can you download [Zed Preview](https://zed.dev/releases/preview) and add this to your Keymap JSON?
> 
>   {
>     "context": "Dock || Workspace || Terminal || OutlinePanel || ProjectPanel || CollabPanel || (Editor && mode == auto_height)",
>     "bindings": { "escape": "editor::ToggleFocus" }
>   },

Yes, this seems to work the way I expect (tapping esc repeatedly ensures that focus remains on the editor). That instantly makes zed much *much* more usable using only the keyboard. 

Thanks!
